src/FakeItEasy.Analyzer/NonVirtualSetupAnalyzer.cs

```diff
@@ -0,0 +1,79 @@
+﻿namespace FakeItEasy.Analyzer
+{
+    using System;
+    using System.Collections.Immutable;
+    using System.Linq;
+    using Microsoft.CodeAnalysis;
+    using Microsoft.CodeAnalysis.CSharp;
+    using Microsoft.CodeAnalysis.CSharp.Syntax;
+    using Microsoft.CodeAnalysis.Diagnostics;
+
+    [DiagnosticAnalyzer(LanguageNames.CSharp)]
+    class NonVirtualSetupAnalyzer:DiagnosticAnalyzer
+    {
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics { get; } = CreateSupportedDiagnostics();
+
+        public override void Initialize(AnalysisContext context)
+        {
+            if (context == null)
+            {
+                throw new ArgumentNullException(nameof(context));
+            }
+
+            context.RegisterSyntaxNodeAction(AnalyzeCall, SyntaxKind.InvocationExpression);
+        }
+
+        private static void AnalyzeCall(SyntaxNodeAnalysisContext context)
+        {
+            var invocationExpression = (InvocationExpressionSyntax)context.Node;
+
+            if (IsInterface(context, invocationExpression))
+                return;
+
+            if (NotSetupCall(invocationExpression))
+                return;
+
+            var symbolInfo = context.SemanticModel.GetSymbolInfo(context.Node);
+
+            if (!symbolInfo.Symbol.IsVirtual)
+            {
+                var location = invocationExpression.GetLocation();
+
+                Diagnostic diagnostic = Diagnostic.Create(DiagnosticDefinitions.NonVirtualSetup, location, symbolInfo.Symbol.Name);
+                context.ReportDiagnostic(diagnostic);
+            }
+        }
+
+        private static bool NotSetupCall(InvocationExpressionSyntax invocationExpression)
+        {
+            var isSetupInvocation = invocationExpression
+                                    .Ancestors()
+                                    .OfType<InvocationExpressionSyntax>()
+                                    .Take(1)
+                                    .Single();

```

Comment:
`.Take(1).Single()` is probably better written as `First()`, but in this case, I'd be inclined to go with `FirstOrDefault()` and return `true` if the result is `null` - as written, we're throwing a lot of unnecessary exceptions whenever there is no `InvocationExpressionSyntax` in `Ancestors`, and extra exceptions are at least wasteful, and can also interfere with usage of the analyzer, as seen in #731.